### ISSUE TITLE
Consider adding API for enumerating all visible Inputs in PVW and/or PGM

### ISSUE THREAD

**Issue by Lange** [nrkno/tv-automation-atem-connection#54]
* **I'm submitting a ...**
[ ] bug report
[x] feature request (I'll code it and open a PR if we think its a good idea)
[ ] question about the decisions made in the repository
[ ] question about how to use this project

* **Summary**
A common task is writing code which reacts to which Inputs are present in PGM and/or PVW. This is harder than it seems at first glance, because we can't simply use the `previewInput`/`programInput` property of the ME state object, because that wouldn't account for keyers and SuperSources. To figure out what Inputs are actually displaing in PVW/PGM right now, we have to loop through all keyers and SuperSources.

This is a common enough task that I think it should be provided by this library. This is my first thought at what this API would look like:

```ts
atem.on('stateChanged', state => {
    const me = state.video.getMe(0);

    // This includes the following Inputs:
    // 1. The current `programInput`
    // 2. If `programInput` is a SuperSource, then it also includes all inputs being used by that SS (recursively)
    // 3. Any inputs currently being shown as DSKs or USKs
    console.log(me.getVisibleInputs('program')); // => logs an array of numbers

    // Same as above but for PVW instead of PGM.
    console.log(me.getVisibleInputs('preview')); // => logs an array of numbers
});
```

I'm not completely married to this API design, and am very open to feedback and suggestions.

cc @baltedewit and @Julusian for their thoughts

**Comment by Julusian** [comment]
I think having this in the library would make sense.

In your checks, you didnt account for if ME2 is referenced anywhere by ME1. 

I don't think it should live on the ME object though, as the checks need to be able to access all of the state.
Perhaps a loose/importable helper function 
```
function listVisibleInputs(state: AtemState, mode: 'program' | 'preview', me = 0): number[]
```

Some questions:
* Do we include any sources used in an active/inactive dip/stinger/dve transition?
* Does program contain all of preview while mid transition?

**Comment by Lange** [comment]
I agree with your points.

> Do we include any sources used in an active/inactive dip/stinger/dve transition?

I think including sources being actively used in a transition makes sense. I guess I'm not sure what you mean by "inactive", or why inactive things would show up as PVW or PGM.

> Does program contain all of preview while mid transition?

I think it should, because in my mind, "PGM" means "on the air in any capacity", which is true for the entirety of the transition. Y'all are more experienced than me though, so if that's not what the industry expects then I'm fine using a different heuristic.

**Comment by baltedewit** [comment]
Great idea, love it. I think this would be perfect as a method on the Atem class as part of the high level API. I don't think it would be useful to run this on a state that isn't the current state so passing the state as a parameter might be redundant.

I think we should attempt to keep this in line with tally output as displayed on the internal multiviewer.

**Comment by Lange** [comment]
> I think we should attempt to keep this in line with tally output as displayed on the internal multiviewer.

Yes, you're totally right, that is what users will expect. Is this well-documented anywhere? If not, my ATEM is already set up in a rack in my living room so I can put in the time to figure that out.

**Comment by Lange** [comment]
Findings:
- Everything involved in a transition is marked as being in PGM for the entire duration of that transition.
    - This inlcudes the current PVW and PGM sources, as well as _any_ source involved in the transition via a DIP, WIPE, STRING, or DVE transition. This _also_ includes both the key _and_ fill of things like the DVE transition.
- Yes, all items of the SuperSource show up as PGM while the SS is in PGM.
    - This recurses down into other MEs if they are one of the SS boxes.
- Yes, taking one of the secondary MEs into ME1 as PGM will result in that secondary ME's current PGM source(s) being marked as being in ME1 PGM.

In short: the multiviewer and tally seem to exclusively care about what ME1 is doing, and will recurse through SuperSources and nested MEs all the way to the leaf nodes to figure out what all is in PVW and/or PGM. And during a transition, everything involved in that transition is marked as PGM for the entirety of said transition.

I think this is all I need to know, and I can begin implementing this algorithm now.

**Comment by Lange** [comment]
I'm making progress, but something I think we're missing is a lookup table of what input numbers correspond to what type of source. For instance, if my code sees that input `123` is in PGM, it needs to know if that ID corresponds to a normal input, supersource, ME, etc. Only then can it properly recurse the tree.
 
I thought I remembered us having something that did this mapping, but I can't find it. Does that only exist in LibAtem, and do I need to port it over here? @Julusian

**Comment by Julusian** [comment]
I was just about to reply with InternalPortType :)

**Comment by Julusian** [comment]
> I think including sources being actively used in a transition makes sense. I guess I'm not sure what you mean by "inactive", or why inactive things would show up as PVW or PGM.

By inactive I meant transition properties for the next transition. I'm not sure why they would show up though. Probably best they dont.

> @baltedewit I think this would be perfect as a method on the Atem class

Yeah, I forget that is somewhere we can put things. I can see that having it separate might be useful in future for something tsr/timeline (ie checking what we are about to do), but lets leave that until it is needed.

Re findings, that sounds like what I remember experiencing with an atem, and is what I would expect. Please make sure we are able to run the algorithm starting from ME2, as that would be really useful when each ME is used in isolation.

**Comment by Lange** [comment]
> Please make sure we are able to run the algorithm starting from ME2, as that would be really useful when each ME is used in isolation.

Will do.

**Comment by Lange** [comment]
Oof okay, hit a new problem:

Say I see an Input with ID `10020`. I can figure out that its an `MEOutput` via `InternalPortType`, but I don't know how to programmatically figure out _which_ ME it corresponds to without [hard coding a map like LibAtem has](https://github.com/LibAtem/LibAtem/blob/7e50edd41dc1ecd086a235cad494a295182e5ecc/LibAtem/Common/VideoSource.cs#L258-L261). Any thoughts?

**Comment by Julusian** [comment]
This has been merged :)
